lib/trace/core.js

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var events = require('events');
+var util = require('util');
+
+var _ = require('underscore');
+
+var tracerConfigMap = require('./config');
+
+// resolves and returns a tracer config object, with
+// extEvent, sourceEmitter, sourceEvent
+function resolveTracerConfig(ringpop, extEvent) {
+    var tracerConfig = tracerConfigMap[extEvent];
+    var emitterPath = tracerConfig.sourcePath;
+
+    // XXX subtle but important -- if we resolve here, then the
+    // underlying object can never change. if we do latent resolution,
+    // there may be a resource leak when we try to remove listener
+    // from something that is out of our reachable scope.
+    var emitter = getIn(ringpop, emitterPath.slice(0, -1));
+    var event = _.last(emitterPath);
+
+    return _.extend(
+        {},
+        tracerConfig,
+        {
+            extEvent: extEvent,
+            sourceEmitter: emitter,
+            sourceEvent: event,
+        }
+    );
+}
+
+function getIn(obj, path) {
+    for (var i = 0;
+         (typeof obj === 'object' || Array.isArray(obj)) && i < path.length;
+         ++i) {
+        obj = obj[path[i]];
+    }
+    return obj;
+}
+
+
+
+// base for sink-specific tracers
+function Tracer(ringpop, config, opts) {
+    if (!(this instanceof Tracer)) {
+        return new Tracer(opts);
+    }
+
+    this.ringpop = ringpop;
+    _.extend(this, config);
+    this.opts = opts;
+    this.listener = null;
+}
+util.inherits(Tracer, events.EventEmitter);
+
+Tracer.validate = function validate(opts) {
+    return (
+        opts &&
+        opts.sink
+    );
+};
+
+Tracer.prototype.validate = function validate() {
```

Comment:
This was unclear. The class-level is now renamed to validateOptions(), since the intent to pre-check opts during subscription handling. The instance-levels check with instances.